contracts/optimizedForceMove.sol

```diff
@@ -0,0 +1,234 @@
+                keccak256(
+                    abi.encodePacked(
+                        fixedPart.challengeDuration,
+                        fixedPart.appDefinition,
+                        variableParts[i].appData
+                    )
+                ),
+                keccak256(abi.encode(variableParts[i].outcome))
+            );
+            stateHashes[i] = keccak256(abi.encode(state));
+            require(
+                _validTransition(largestTurnNum + i - m, variableParts[i], variableParts[i + 1]),
+                'Invalid Transition'
+            );
+        }
+
+        // check the supplied states are supported by n signatures
+        require(
+            _validSignatures(
+                largestTurnNum,
+                fixedPart.participants,
+                stateHashes,
+                sigs,
+                whoSignedWhat
+            ),
+            'Invalid signature'
+        );
+
+        // check that the forceMove is signed by a participant and store their address
+        bytes32 msgHash = keccak256(
+            abi.encode(
+                largestTurnNum,
+                channelId,
+                'forceMove' // Express statement of intent to forceMove this channel at this turnNum
+            )
+        );
+        address challenger = _recoverSigner(
+            msgHash,
+            challengerSig.v,
+            challengerSig.r,
+            challengerSig.s
+        );
+        require(
+            _isAddressInArray(challenger, fixedPart.participants),
+            'Challenger is not a participant'
+        );
+
+        // ------------
+        // EFFECTS
+        // ------------
+
+        channelStorage = ChannelStorage(
+            largestTurnNum,
+            now + fixedPart.challengeDuration,
+            stateHashes[m],
+            challenger,
+            keccak256(abi.encode(variableParts[m].outcome))
+        );
+
+        channelStorageHashes[channelId] = keccak256(abi.encode(channelStorage));
+    }
+    // Internal methods:
+
+    function _isAddressInArray(address suspect, address[] memory addresses)
+        internal
+        pure
+        returns (bool)
+    {
+        for (uint256 i = 0; i < addresses.length; i++) {
+            if (suspect == addresses[i]) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function _validSignatures(
+        uint256 largestTurnNum,
+        address[] memory participants,
+        bytes32[] memory stateHashes,
+        Signature[] memory sigs,
+        uint8[] memory whoSignedWhat // whoSignedWhat[i] is the index of the state in stateHashes that was signed by participants[i]
+    ) internal pure returns (bool) {
+        uint256 nParticipants = participants.length;
+        uint256 nStates = stateHashes.length;
+
+        require(
+            whoSignedWhat.length == nParticipants,
+            '_validSignatures: whoSignedWhat must be the same length as participants'
+        );
```

Comment:
Maybe this should be part of `_acceptableWhoSignedWhat`

Reply:
:+1: